readseeker.go

```diff
@@ -0,0 +1,153 @@
+		ip.pos += delta
+		ip.curChunkOffset += delta
+		return ip.pos, nil
+	}
+
+	// General case: Bisect
+	chunks := ip.Index.Chunks
+	newChunkIdx = sort.Search(len(chunks), func(i int) bool { return int64(chunks[i].Start) >= newPos })
+	if newChunkIdx >= len(chunks) { // function was not true for any chunk -- meaning we're running off the end
+		newChunkIdx = len(chunks) - 1
+	}
+	newChunk := ip.Index.Chunks[newChunkIdx]
+	newChunkOffset = newPos - int64(newChunk.Start)
+
+	if newPos < int64(newChunk.Start) {
+		return ip.pos, fmt.Errorf("Seek found chunk beginning at position %v, desired position is %v", newChunk.Start, newPos)
+	}
+	if newPos > int64(newChunk.Start+newChunk.Size) {
+		return ip.pos, fmt.Errorf("Seek found chunk ending at position %v, desired position is %v", newChunk.Start+newChunk.Size, newPos)
+	}
+
+	// Only invalidate cache if new chunk is different from old one (avoid re-decompressing all-0 regions)
+	if newChunk.ID != ip.curChunkID {
+		ip.curChunk = nil // next read attempt will call loadChunk()
+	}
+	// BELOW HERE, WE HAVE UPDATED THE DATA AND MUST NOT ERROR
+	ip.curChunkIdx = newChunkIdx
+	ip.curChunkID = newChunk.ID
+	ip.curChunkOffset = newChunkOffset
+	ip.pos = newPos
+	return newPos, err
+}
+
+func (ip *IndexPos) loadChunk() (err error) {
+	var compressedChunk []byte
+	var decompressedChunk []byte
+
+	compressedChunk, err = ip.Store.GetChunk(ip.curChunkID)
+	if err != nil {
+		return err
+	}
+
+	decompressedChunk, err = Decompress(nil, compressedChunk)
+	if err != nil {
+		return err
+	}
+
+	ip.curChunk = decompressedChunk
+	return nil
+}
+
+func (ip *IndexPos) Seek(offset int64, whence int) (int64, error) {
+	var newPos int64
+	var err error
+	switch whence {
+	case io.SeekStart:
+		newPos = offset
+	case io.SeekCurrent:
+		newPos = ip.pos + offset
+	case io.SeekEnd:
+		newPos = ip.Length - offset
+	default:
+		return ip.pos, fmt.Errorf("invalid whence")
+	}
+	if newPos < 0 {
+		return ip.pos, fmt.Errorf("unable to seek before start of file")
+	}
+	newOffset, err := ip.findOffset(newPos)
+	if err == nil && newPos > ip.Length {
+		err = io.EOF
+	}
+	return newOffset, err
+}
+
+func (ip *IndexPos) Read(p []byte) (n int, err error) {
+	var totalCopiedBytes int
+	remainingBytes := p[:]
+	if ip.pos == ip.Length { // if initially called when already at the end, return EOF
+		return 0, io.EOF
+	}
+	for len(remainingBytes) > 0 {
+		if len(ip.curChunk) == 0 {
+			err = ip.loadChunk()
+			if err != nil {
+				break
+			}
+		}
+		chunkRemainingBytes := ip.curChunk[ip.curChunkOffset:len(ip.curChunk)]
+		if len(chunkRemainingBytes) == 0 && ip.curChunkIdx == (len(ip.Index.Chunks)-1) {
+			break // if running into the end after successful read, return a short read
+		}
+		copiedBytes := copy(remainingBytes, chunkRemainingBytes)
+		remainingBytes = remainingBytes[copiedBytes:]
+		totalCopiedBytes += copiedBytes
+		_, err := ip.Seek(int64(copiedBytes), io.SeekCurrent)
```

Comment:
Pretty sure this `err` masks the one returned at the end of the function. So an error by Seek() isn't going to be returned. Removing `:` should do the trick.